### PR TITLE
[WIP] i18n

### PR DESCRIPTION
**Pull request by dachary**
## Status

Work in progress

## Description of Changes

i18n of source and journalist interface. This pull request is for reference and is not meant to be merged as is. Commits are proposed as separated pulll requests. Note that the [order of the commits as presented by github is not what you would get with git log](https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/). When viewed with git log the commits are grouped for clarity.

## Active pull requests waiting to be merged

These are the pull requests created from various commits in this branch.There should not be more than a few otherwise it will become difficult to manage.

* https://github.com/freedomofpress/securedrop/pull/2343
* https://github.com/freedomofpress/securedrop/pull/2344
* https://github.com/freedomofpress/securedrop/pull/2345
* https://github.com/freedomofpress/securedrop/pull/2346
* https://github.com/freedomofpress/securedrop/pull/2347
* https://github.com/freedomofpress/securedrop/pull/2348
* https://github.com/freedomofpress/securedrop/pull/2349

### CODE REVIEW

**Comment by dachary** [comment]
<pre>
[gw5] FAILED testinfra/functional/test_tor_interfaces.py::test_www[docker://apptestclient-site0] 
E       AssertionError: assert '500' == '200'
E         - 500
E         + 200
</pre>
does not seem related.

**Comment by dachary** [comment]
<pre>
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to validate the SSL certificate for launchpad.net:443. Make sure your managed systems have a valid CA certificate installed. If the website serving the url uses SNI you need python >= 2.7.9 on your managed machine or you can install the `urllib3`, `pyopenssl`, `ndg-httpsclient`, and `pyasn1` python modules to perform SNI verification in python >= 2.6. You can use validate_certs=False if you do not need to confirm the servers identity but this is unsafe and not recommended. Paths checked for this platform: /etc/ssl/certs, /etc/pki/ca-trust/extracted/pem, /etc/pki/tls/certs, /usr/share/ca-certificates/cacert.org, /etc/ansible"}
</pre>
interesting transient error, let's try again

**Comment by dachary** [comment]
<pre>
[gw5] FAILED testinfra/functional/test_tor_interfaces.py::test_www[docker://apptestclient-site0] 
</pre>
again. There is no indication about what's in the logs though, therefore no way to know what the error is about. @msheiny do you have a suggestion on how to get more information [from this run](https://circleci.com/gh/freedomofpress/securedrop/1170?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) ?

**Comment by dachary** [comment]
<pre>
>       raise TimeoutException(message, screen, stacktrace)
E       TimeoutException: Message: Timed out waiting for confirmation popup.
</pre>
I assume [travis was a little too slow](https://travis-ci.org/freedomofpress/securedrop/builds/265482275?utm_source=github_status&utm_medium=notification) and timeout should be increased to reduce the chances of this happening.

**Comment by dachary** [comment]
The [travis error](https://travis-ci.org/freedomofpress/securedrop/builds/266956968?utm_source=github_status&utm_medium=notification) cannot be reproduced localy, re-spawning to verify this is a transient error.

<pre>
self = <test_journalist.TestJournalistLayout instance at 0x7fefd0394ef0>
    def _journalist_selects_first_doc(self):
>       self.driver.find_elements_by_name('doc_names_selected')[0].click()
E       IndexError: list index out of range
</pre>

**Comment by dachary** [comment]
False negative [on circle-ci](https://circleci.com/gh/freedomofpress/securedrop/1262?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) due to failed AWS provisioning.
<pre>
fatal: [mon-staging -> localhost]: FAILED! => {"changed": false, "elapsed": 601, "failed": true, "msg": "Timeout when waiting for search string OpenSSH in ec2-13-56-12-1.us-west-1.compute.amazonaws.com:22"}
	to retry, use: --limit @/root/sd/devops/playbooks/reboot_and_wait.retry
</pre>

**Comment by dachary** [comment]
<pre>
self = <tests.test_journalist.TestJournalistApp testMethod=test_delete_source_deletes_submissions>
    def tearDown(self):
>       utils.env.teardown()
tests/test_journalist.py:45: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/utils/env.py:71: in teardown
    shutil.rmtree(config.SECUREDROP_DATA_ROOT)
/usr/lib/python2.7/shutil.py:247: in rmtree
    rmtree(fullname, ignore_errors, onerror)
/usr/lib/python2.7/shutil.py:256: in rmtree
    onerror(os.rmdir, path, sys.exc_info())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
path = '/tmp/securedrop/keys', ignore_errors = False
onerror = <function onerror at 0x7fb392fb7410>
    def rmtree(path, ignore_errors=False, onerror=None):
..
       try:
>           os.rmdir(path)
E           OSError: [Errno 39] Directory not empty: '/tmp/securedrop/keys'
</pre>
**sigh** it's still here. Transient false negative.

**Comment by dachary** [comment]
rebased & repushed, including https://github.com/freedomofpress/securedrop/pull/2184 to get coverage until it is merged

**Comment by dachary** [comment]
CircleCI failed because AWS hosts provisioning failed (see https://circleci.com/gh/freedomofpress/securedrop/1414?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link for details)

**Comment by dachary** [comment]
rebased & repushed. A dozen conflicts from https://github.com/freedomofpress/securedrop/pull/2079.

**Comment by dachary** [comment]
travis had connectivity issues, re-pushing
<pre>
    socket.error: [Errno 104] Connection reset by peer
</pre>

**Comment by dachary** [comment]
rebased & resolved conflicts introduced by https://github.com/freedomofpress/securedrop/pull/2227
repushed !

**Comment by dachary** [comment]
rebased & resolved conflicts introduced by #2235
repushed !

**Comment by dachary** [comment]
https://circleci.com/gh/freedomofpress/securedrop/1778?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

<pre>
Timeout when waiting for ec2-54-183-192-4.us-west-1.compute.amazonaws.com:22
</pre>

**Comment by dachary** [comment]
https://circleci.com/gh/freedomofpress/securedrop/1780?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

<pre>
fatal: [app-staging]: FAILED! => {"changed": true, "failed": true, "rc": 1, "stderr": "Connection to ec2-54-193-101-30.us-west-1.compute.amazonaws.com closed.\r\n", "
</pre>

**Comment by msheiny** [comment]
:( Okay -- So #2218 is fix these connection issues at the very beginning of the play but I dont have a solution for this specific issue yet. I made a new ticket though to solve this particular bug #2253

**Comment by dachary** [comment]
@msheiny it is not an actual error, my bad. The real error is above and entirely my fault:
<pre>
such file or directory: '/var/www/securedrop/tests/i18n/source.desktop.in'
</pre>

**Comment by redshiftzero** [comment]
Just adding a note here so I (or whoever reviews the remainder of this): We're in a bit of a tricky situation with `DEFAULT_LOCALE` (related: #1966). In #2197 we added this to `config.py` but it requires a playbook run to be added to existing installs, so unless admins run the playbooks (which we know they don't often do), we _cannot expect this variable to be present_ in existing installs. So the app code should assume `en_US` as a default if `config.DEFAULT_LOCALE` does not exist.

File: securedrop/i18n.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    else:
+        if 'locale' in session:
+            locale = session['locale']
+        else:
+            locale = core.negotiate_locale(accept_languages, LOCALES)
+
+    if locale:
+        return locale
+    else:
+        return getattr(config, 'DEFAULT_LOCALE')
```
Comment (dachary) [inline securedrop/i18n.py]: return getattr(config, 'DEFAULT_LOCALE', 'en_US')

**Comment by dachary** [comment]
> en_US as a default if config.DEFAULT_LOCALE does not exist.

good catch ! https://github.com/freedomofpress/securedrop/pull/2116/files#r137906995

**Comment by dachary** [comment]
rebased & repushed, conflicts introduced by:

* https://github.com/freedomofpress/securedrop/pull/2227
* https://github.com/freedomofpress/securedrop/pull/2279
* https://github.com/freedomofpress/securedrop/pull/2265

File: securedrop/i18n.py (hunk 1 above)
Comment (dachary) [reply securedrop/i18n.py]: done

**Comment by dachary** [comment]
A number of page-layout screenshot failed at https://travis-ci.org/freedomofpress/securedrop/builds/276838839?utm_source=github_status&utm_medium=notification relaunching to verify this is not a transient error

**Comment by dachary** [comment]
pip install broke on travis https://travis-ci.org/freedomofpress/securedrop/builds/277684950?utm_source=github_status&utm_medium=notification

<pre>
'Connection reset by peer')\", error(104, 'Connection reset by peer')
</pre>

**Comment by dachary** [comment]
The remaining commits from this long standing PR are in two pull requests. The backlog is empty and i18n no longer requires special care, it is part of the usual development process.

**Pull request closed without merging**